models/Category.php

```diff
@@ -154,40 +187,69 @@ public static function getMenuTypeInfo($type)
     public static function resolveMenuItem($item, $url, $theme)
     {
         $structure = [];
-        $category  = new Category();
 
-        if ( ! $pageUrl = GeneralSettings::get('category_page')) {
-            throw new InvalidArgumentException(
-                'Mall: Please select a category page via the backend settings.'
-            );
-        }
+        if($item->type == 'mall-category') {
+            $category = self::find($item->reference);
+            if (!$category) {
+                return;
+            }
 
-        $iterator = function ($items, $baseUrl = '') use (&$iterator, &$structure, $pageUrl, $url) {
-            $branch = [];
+            $structure = self::getMenuItem($category, $url);
 
-            $controller = new Controller();
-            foreach ($items as $item) {
-                $entryUrl               = $controller->pageUrl($pageUrl, ['slug' => $item->nestedSlug]);
-                $branchItem             = [];
-                $branchItem['url']      = $entryUrl;
-                $branchItem['isActive'] = $entryUrl === $url;
-                $branchItem['title']    = $item->name;
+        } elseif ($item->type == 'all-mall-categories') {
+            $category  = new Category();
 
-                if ($item->children) {
-                    $branchItem['items'] = $iterator($item->children, $item->slug);
-                }
+            $iterator = function ($items, $baseUrl = '') use (&$iterator, &$structure, $url) {
+                $branch = [];
 
-                $branch[] = $branchItem;
-            }
+                foreach ($items as $item) {
+                    $branchItem = self::getMenuItem($item, $url);
 
-            return $branch;
-        };
+                    if ($item->children) {
+                        $branchItem['items'] = $iterator($item->children, $item->slug);
+                    }
 
-        $structure['items'] = $iterator($category->getEagerRoot());
+                    $branch[] = $branchItem;
+                }
+
+                return $branch;
+            };
+
+            $structure['items'] = $iterator($category->getEagerRoot());
+        }
 
         return $structure;
     }
 
+    /**
+     * Creates a single menu item result array
+     *
+     * @param $item Category
+     * @param $url string
+     *
+     * @return array
+     */
+    protected static function getMenuItem($item, $url)
+    {
+        if ( ! $pageUrl = GeneralSettings::get('category_page')) {
```

Comment:
Why did you move this guard here? Now it gets called once for every menu item. Can you move this back to `resolveMenuItem`?

Reply:
No, sounds good to me!